HPCC-29029 Add guidelines for creating pull requests

ghalliday opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

ghalliday committed:
HPCC-29029 Add guidelines for creating pull requests

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday commented:
All - for discussion.  I've split the code review document in two.  One aimed at people submitting prs, and the other at reviewers.  All comments/missing ideas/etc very welcome.

richardkchapman reviewed (changes requested):
Couple of typos but looks generally sensible

jakesmith reviewed (commented):
@ghalliday - looks good, a few minor comments

GordonSmith commented on devdoc/CodeReviews.md:
```diff
@@ -90,5 +103,5 @@ documentation: | This change may have an impact on documentation | Make sure cha
 The comments should always be constructive.  The reviewer should have a reason for each of them, and be able to articulate the reason in the comment or when asked.  "I wouldn't have done it like that" is not a good enough on its own!
 
-Similarly there is a difference in opinion within the team on some style issues - e.g. standard libraries or jlib, inline or out of line functions, nested or non-nested classes.  Reviews should try and avoid commenting on these unless there is a clear reason why they are significant (functionality, efficiency, compile time) and if so spell it out.  Code reviewers should discuss any style issues t ...
+Similarly there is a difference in opinion within the team on some style issues - e.g. standard libraries or jlib, inline or out of line functions, nested or non-nested classes.  Reviews should try and avoid commenting on these unless there is a clear reason why they are significant (functionality, efficiency, compile time) and if so spell it out.  Code reviewers should discuss any style issues t ...
```
Relative URLs are preferable

GordonSmith commented:
@ghalliday to add this document to the landing page LHS menu, you can edit this file:  https://github.com/hpcc-systems/HPCC-Platform/blob/master/devdoc/.vitepress/config.js
In general I would keep that in sync with:  https://github.com/hpcc-systems/HPCC-Platform/blob/master/devdoc/README.md

cloLN commented on devdoc/CodeSubmissions.md:
```diff
@@ -0,0 +1,71 @@
+## Pull requests
+There are a few things that should be considered when creating a PR to increase the likelihood that they can be accepted quickly.
+
+* Write a good commit message\
```
commit messages should reference a specific JIRA whenever possible,  Even if it's an update due to code review, it should reference the issue number or at the very least be more descriptive.

ghalliday replied:
I'm not sure that follow on commits need to the issue number - unless they are going to be preserved when the change is merged.  If they will be squashed then it is ok to say "changes following review"

ghalliday commented:
@GordonSmith added links in readme and config.js.  Fixed the broken links from readme (requires the .md extension).
Added Richard's recent roxie doc.

GordonSmith reviewed (approved):
Looks good from the markdown -> doc site generation point of view.

@ghalliday if you want to render it while editing you can:
```
npm install
npm run docs-dev
```

stuartort commented on devdoc/CodeSubmissions.md:
```diff
@@ -0,0 +1,76 @@
+Contributors should use the github reviewers section on the PR to request reviews.  After a contributor has pushed a set of changes in response to a review, they should refresh the github review status, so the users are notified it is ready for re-review.  When the review is complete, a person with responsibility for merging changes to that part of the system should be added as a reviewer (or ref ...
+
+Reviewers should check for PRs that are ready for their review via github's webpage (filter "review-requested:\<reviewer-id>") or via the github CLI (e.g. gh pr status).  Contributors should similarly ensure they stay up to date with any comments on requests for change on their submissions.
+
```
+ Reviewers should make an try to find all issues during the first review.  Often there are issues found by a reviewer, fixed by the contributor and then additional issues are reported by the reviewer (unrelated to the first set of changes)
+
+ If PR requires multiple reviewers, the contributor should communicate with reviewers for the parts that need multiple people to review.  This can help reduce the times where a reviewer makes comments, the contributor addresses that issue and a different reviewer objects to the change.  It would be better if the contributor can request all reviewers to comment before addressing the issue
+
+When doing a code review, the reviewer should be careful not to request additional features that are outside the scope of the initial jira (i.e. this feature would be better if you add these 2 new bits of functionality).  This should be a new jira request and not hold up the current PR

stuartort commented:
In the CodeReviews.md file, I feel it would be good to add to like the following (I may be violating the last point since this is an area that isn't under review at this time )
+ Reviewers should make an try to find all issues during the first review.  Often there are issues found by a reviewer, fixed by the contributor and then additional issues are reported by the reviewer (unrelated to the first set of changes)
+
+ If PR requires multiple reviewers, the contributor should communicate with reviewers for the parts that need multiple people to review.  This can help reduce the times where a reviewer makes comments, the contributor addresses that issue and a different reviewer objects to the change.  It would be better if the contributor can request all reviewers to comment before addressing the issue
+
+ When doing a code review, the reviewer should be careful not to request additional features that are outside the scope of the initial jira (i.e. this feature would be better if you add these 2 new bits of functionality).  This should be a new jira request and not hold up the current PR

ghalliday commented:
> In the CodeReviews.md file, I feel it would be good to add to like the following (I may be violating the last point since this is an area that isn't under review at this time )
> 
> * Reviewers should make an try to find all issues during the first review.  Often there are issues found by a reviewer, fixed by the contributor and then additional issues are reported by the reviewer (unrelated to the first set of changes)
It should be an intention, but I don't think we can require this - I often find additional problems on a second review.  Sometimes it is because there are serious problems in the original PR, and then I comment that not all issues will have been reported.  Other times it is because you think of things a second time - possibly because you are now more familiar with the code.

> * If PR requires multiple reviewers, the contributor should communicate with reviewers for the parts that need multiple people to review.  This can help reduce the times where a reviewer makes comments, the contributor addresses that issue and a different reviewer objects to the change.  It would be better if the contributor can request all reviewers to comment before addressing the issue
Not sure how to address this one.  If that ever occurs the two reviewers should discuss how to avoid it in the future.

> * When doing a code review, the reviewer should be careful not to request additional features that are outside the scope of the initial jira (i.e. this feature would be better if you add these 2 new bits of functionality).  This should be a new jira request and not hold up the current PR
Yes, that's a good point.  I'll include something like that.
